### DISCUSSION TITLE
Regarding the future of TShock versions 4, 5, 6, and 7

### DISCUSSION THREAD

**Discussion by hakusaro** [Pryaxis/TShock#2673] [Announcements]
Hello!

TShock has always been a community maintained project. We give maintainers commit access to TShock and welcome them into Pryaxis and the TShock Team when they've contributed good changes and have good opinions about various things. All of the contributors to TShock have been added because we love their work, and want to encourage them. TShock is nothing without its contributors.

In light of this, we've had an unspoken policy that closely mirrors the [WP:BE BOLD](https://en.wikipedia.org/wiki/Wikipedia:Be_bold) policy on Wikipedia. While the entire thing is worth reading for some people, I'll quote an important point here:

> Fix it yourself instead of just talking about it. If you notice an unambiguous error or problem that any reasonable person would recommend fixing, the best course of action may be to be bold and fix it yourself rather than bringing it to someone's attention in the form of a comment or complaint. In the time it takes to write about the problem, you could instead improve [it].

In light of this, we encourage dramatic changes, and we encourage people to take responsibility for changes they want to make.

There has been a lot of activity behind the scenes, and some of this activity is starting to bear fruit.

## OTAPI 3

First and foremost, I'd like to extend my thanks and admiration to @SignatureBeef, the creator of TDSM. Back in the day, TDSM and TShock "competed" with tMod. Luke has been pouring his soul into this game, and has built the most advanced mod framework and development pipeline for us to work on, and has tirelessly ported TShock to it. The Open Terraria API is what enabled TShock to live on beyond our natural capabilities as a decompiled-and-recompiled server mod. In OTAPI 2, what TShock 4 is based on, TShock's old API, the Terraria Server API, from bladecoding, is shimmed on top of OTAPI.

In the upcoming version of TShock, TShock 5, TSAPI is renamed to just "TSAPI", or the Terraria Scaffold API. This loads TShock against the new version of OTAPI. Version 3 introduces many changes, most notably making it far easier to patch TShock for future updates. In addition, it brings us out of the stone age. TShock 5 and TSAPI 5 will be published on nuget, and consumable using nuget-friendly workflows for developers. Better still, it'll support the latest version of .NET 6.

Again, this is all the work of Luke, aka, @SignatureBeef, an incredibly kind and lovely person.

## TShock 6

Building on the work from @SignatureBeef on OTAPI 3, @QuiCM and @SignatureBeef have been working hard to rebuild core TShock internal structures around dependency injection. This will make plugin development _more sane_ in the long run. It will also require work from plugin developers to port their work to the new system. Unfortunately, Re-Logic's timing with 1.4.4 couldn't have been worse: we planned to build TShock 5 on OTAPI 3 with these changes, but have postponed them because they won't be done for a few more weeks. TShock 6 will introduce several breaking plugin changes, and require developers to update. However, this will make the TShock codebase much more well-supported for the future.

This time, my thanks extends to Chris, aka @QuiCM, a lovely individual who took the liberty of improving the codebase and working with Luke, aka @SignatureBeef on it. You two really are hitting it out of the park!

## TShock 7

TShock 7 is the version after TShock 5, and 6. It will add support for .NET 7, the newest version of the .NET framework. This will likely occur sometime in the next several weeks, after TShock 6.

## For plugin developers

We expect to have a lot of documentation on migration paths soon. Rest assured, we will document what you need to do.

## Conclusion

I absolutely love @SignatureBeef and @QuiCM. Thanks for all of your work on the latest updates!

**Comment by IraJacobs** [comment]
Keep up the great work, all of you!  Very much appreciated!

**Comment by bippity** [comment]
WHOOOO THANK YOU BOYOS!
